storage/src/main/java/com/tencent/rss/storage/handler/impl/DataSkippableReadHandler.java

```diff
@@ -0,0 +1,94 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.tencent.rss.storage.handler.impl;
+
+import java.util.List;
+
+import com.google.common.collect.Lists;
+import org.roaringbitmap.longlong.Roaring64NavigableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.tencent.rss.common.ShuffleDataResult;
+import com.tencent.rss.common.ShuffleDataSegment;
+import com.tencent.rss.common.ShuffleIndexResult;
+import com.tencent.rss.common.util.RssUtils;
+
+public abstract class DataSkippableReadHandler extends AbstractClientReadHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(DataSkippableReadHandler.class);
+
+  protected List<ShuffleDataSegment> shuffleDataSegments = Lists.newArrayList();
+  protected int segmentIndex = 0;
+
+  protected Roaring64NavigableMap expectBlockIds;
+  protected Roaring64NavigableMap processBlockIds;
+
+  public DataSkippableReadHandler(
+    String appId,
+    int shuffleId,
+    int partitionId,
+    int readBufferSize,
+    Roaring64NavigableMap expectBlockIds,
+    Roaring64NavigableMap processBlockIds) {
+    this.appId = appId;
+    this.shuffleId = shuffleId;
+    this.partitionId = partitionId;
+    this.readBufferSize = readBufferSize;
+    this.expectBlockIds = expectBlockIds;
+    this.processBlockIds = processBlockIds;
+  }
+
+  protected abstract ShuffleIndexResult readShuffleIndex();
+
+  protected abstract ShuffleDataResult readShuffleData(ShuffleDataSegment segment);
+
+  public ShuffleDataResult readShuffleData() {
+    if (shuffleDataSegments.isEmpty()) {
+      ShuffleIndexResult shuffleIndexResult = readShuffleIndex();
+      if (shuffleIndexResult == null || shuffleIndexResult.isEmpty()) {
+        return null;
+      }
+
+      shuffleDataSegments = RssUtils.transIndexDataToSegments(shuffleIndexResult, readBufferSize);
+    }
+
+    // We should skip unexpected and processed segments when handler is read
+    ShuffleDataResult result = null;
+    while (segmentIndex < shuffleDataSegments.size()) {
+      ShuffleDataSegment segment = shuffleDataSegments.get(segmentIndex);
+      Roaring64NavigableMap blocksOfSegment = Roaring64NavigableMap.bitmapOf();
+      segment.getBufferSegments().forEach(block -> blocksOfSegment.addLong(block.getBlockId()));
+      // skip unexpected blockIds
+      blocksOfSegment.and(expectBlockIds);
+      if (blocksOfSegment.getLongCardinality() > 0) {
+        // skip processed blockIds
+        blocksOfSegment.or(processBlockIds);
```

Comment:
Can we merge `or` and `xor` to one operation? bitmap's performance is worse, do you test it ?

Reply:
or + xor is a fast method to found the item of "1" in blocksOfSegment which are not in processBlockIds.
It is fast because it handles a batch of items by one shot, which "find" only handles each item individually.

By the way, it is a quiz in leetcode.

Reply:
[contain] time is of 10000 items is 3
[or+xor] time is of 10000 items is 0
================
[contain] time is of 20000 items is 2
[or+xor] time is of 20000 items is 0
================
[contain] time is of 30000 items is 3
[or+xor] time is of 30000 items is 0
================
[contain] time is of 40000 items is 3
[or+xor] time is of 40000 items is 0
================
[contain] time is of 50000 items is 4
[or+xor] time is of 50000 items is 0
================
[contain] time is of 60000 items is 4
[or+xor] time is of 60000 items is 0
================
[contain] time is of 70000 items is 7
[or+xor] time is of 70000 items is 4
================
[contain] time is of 80000 items is 7
[or+xor] time is of 80000 items is 0
================
[contain] time is of 90000 items is 7
[or+xor] time is of 90000 items is 0
================
[contain] time is of 100000 items is 8
[or+xor] time is of 100000 items is 1
================

Reply:
public void benchBitMap() {
    for (int i = 10000; i <= 100000; i += 10000) {
      Roaring64NavigableMap A = Roaring64NavigableMap.bitmapOf();
      Roaring64NavigableMap B = Roaring64NavigableMap.bitmapOf();
      for (int j = 0; j < i; j++) {
        if (j % 2 == 0) A.addLong(j);
        if (j % 2 == 1) B.addLong(j);
      }
      long start = System.currentTimeMillis();
      for (int j = 0; j < i; j++) {
        A.contains(j);
      }
      long t1 = System.currentTimeMillis();
      A.or(B);
      A.xor(B);
      long t2 = System.currentTimeMillis();
      System.out.println("[contain] time is of " + i + " items is " + (t1 - start));
      System.out.println("[or+xor] time is of " + i + " items is " + (t2 - t1));
      System.out.println("================");
    }
  }``

Reply:
OK